### PR TITLE
[SPARK-7316] [MLlib] RDD sliding window with step

### PR DESCRIPTION
**Pull request by avulanov**
Implementation of step capability for sliding window function in MLlib's RDD. 

Though one can use current sliding window with step 1 and then filter every Nth window, it will take more time and space (N*data.count times more than needed). For example, below are the results for various windows and steps on 10M data points:

| Window | Step | Time | Windows produced |
| --- | --- | --- | --- |
| 128 | 1 | 6.38 | 9999873 |
| 128 | 10 | 0.9 | 999988 |
| 128 | 100 | 0.41 | 99999 |
| 1024 | 1 | 44.67 | 9998977 |
| 1024 | 10 | 4.74 | 999898 |
| 1024 | 100 | 0.78 | 99990 |

```
import org.apache.spark.mllib.rdd.RDDFunctions._
val rdd = sc.parallelize(1 to 10000000, 10)
rdd.count
val window = 1024
val step = 1
val t = System.nanoTime(); val windows = rdd.sliding(window, step); println(windows.count); println((System.nanoTime() - t) / 1e9)
```

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/RDDFunctions.scala [hunk 1]
```diff
@@ -37,12 +37,12 @@ class RDDFunctions[T: ClassTag](self: RDD[T]) extends Serializable {
    * trigger a Spark job if the parent RDD has more than one partitions and the window size is
    * greater than 1.
    */
-  def sliding(windowSize: Int): RDD[Array[T]] = {
+  def sliding(windowSize: Int, step: Int = 1): RDD[Array[T]] = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/RDDFunctions.scala]: Doesn't this change the API though?

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala [hunk 2]
```diff
@@ -44,15 +44,16 @@ class SlidingRDDPartition[T](val idx: Int, val prev: Partition, val tail: Seq[T]
  * @see [[org.apache.spark.mllib.rdd.RDDFunctions#sliding]]
  */
 private[mllib]
-class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int)
+class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int, val step: Int)
   extends RDD[Array[T]](parent) {
 
-  require(windowSize > 1, s"Window size must be greater than 1, but got $windowSize.")
+  //require(windowSize > 1, s"Window size must be greater than 1, but got $windowSize.")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala]: Why wouldn't this be required anymore?

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/RDDFunctions.scala (hunk 1 above)
Comment (avulanov) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/RDDFunctions.scala]: Should I add a new function instead? Thought there is a default value.

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala (hunk 2 above)
Comment (avulanov) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala]: Sure, fixed.

**Comment by avulanov** [comment]
@srowen @mengxr Added performance comparison https://github.com/apache/spark/pull/5855#issue-72548837

**Comment by andrewor14** [comment]
Hi @avulanov would you mind changing the title to `[SPARK-7316] [MLlib] RDD sliding window with step`? This will help us sort our pull requests properly.

**Comment by avulanov** [comment]
@mengxr Do you plan to add this in 1.5 ?

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala [hunk 3]
```diff
@@ -66,36 +69,54 @@ class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int
...
     } else if (n == 1) {
-      Array(new SlidingRDDPartition[T](0, parentPartitions(0), Seq.empty))
+      Array(new SlidingRDDPartition[T](0, parentPartitions(0), Seq.empty, 0))
     } else {
       val n1 = n - 1
-      val w1 = windowSize - 1
-      // Get the first w1 items of each partition, starting from the second partition.
-      val nextHeads =
-        parent.context.runJob(parent, (iter: Iterator[T]) => iter.take(w1).toArray, 1 until n)
+      // Get partitions sizes
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala]: Keep the original implementation if `step == 1` to save one job.

File: mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala [hunk 4]
```diff
@@ -66,36 +69,54 @@ class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int
...
+      val sizes =
+        parent.context.runJob(parent, (iter: Iterator[T]) => iter.length, 0 until n)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala]: If `step > 1`, in a single Spark job, we can collect the following:

1. the first `w1` elements from each partition (except the first one)
2. the size of each partition

Then we can compute the offset for each partition.

**Comment by avulanov** [comment]
@mengxr Thank you for the review. I replaced two Jobs with one.

**Comment by mengxr** [comment]
@avulanov I sent you a PR at https://github.com/avulanov/spark/pull/3. Since `sliding` is not frequently used, I think simple implementation is preferred over optimization.

**Pull request closed without merging**